### PR TITLE
Warnings

### PR DESCRIPTION
**Pull request by TomAugspurger**
First pass at https://github.com/dask/dask/issues/4142. I've done everything except dask.array, dask.dataframe, and dask.bag. I'll likely split those into separate PRs for ease of review.

Once those are all clean on our CI, we can update our `setup.cfg` to error on warnings (perhaps from specific packages).

### CODE REVIEW

File: dask/tests/test_delayed.py [hunk 1]
```diff
@@ -40,47 +40,47 @@ def __dask_postcompute__(self):
         return tuple, ()
 
 
+@pytest.mark.filterwarnings("ignore:The dask.delayed:UserWarning")
 def test_to_task_dask():
-    with warnings.catch_warnings(record=True):
```
Comment (TomAugspurger) [inline dask/tests/test_delayed.py]: Subtle interaction between pytest and the warnings module, `catch_warnings` just prevents the warnings from being displayed. They still are emitted, and so pytest will still fail with `-Werror`. We would need to either use `warnings.catch_warnings()` with a `warnings.simplefilter`, or `pytest.mark.filterwarnings`.

File: dask/tests/test_sharedict.py [hunk 2]
```diff
@@ -1,15 +1,28 @@
...
+    warnings.simplefilter("ignore", UserWarning)
+    from dask.sharedict import ShareDict
 
 
 a = {'x': 1, 'y': 2}
 b = {'z': 3}
 c = {'w': 2}
 
 
+def test_import_warns():
```
Comment (TomAugspurger) [inline dask/tests/test_sharedict.py]: Hopefully this test is robust to all the import caching that could occur.

Comment (TomAugspurger) [reply dask/tests/test_sharedict.py]: Hmm, seems like it's not robust... I'm just going to assume that we are doing things correctly and not test this.

File: dask/tests/test_delayed.py (hunk 1 above)
Comment (mrocklin) [reply dask/tests/test_delayed.py]: Good to know.  Thanks for the public service announcement.  I'll endeavor to remember this for the future (but I suspect that I'll probably forget too occasionally)

**Comment by mrocklin** [comment]
+1 from me on this.  Thanks @TomAugspurger

**Comment by TomAugspurger** [comment]
Merging this, will followup with additional cleanups.

**Pull request merged**